[OpenXR] Do not duplicate the DMA-BUF file descriptors before importing the external texture

carlosgcampos opened the pull request:
#### 46f0a4c36be821652d1582cbfabe5e51cfee5176
<pre>
[OpenXR] Do not duplicate the DMA-BUF file descriptors before importing the external texture
<a href="https://bugs.webkit.org/show_bug.cgi?id=297891">https://bugs.webkit.org/show_bug.cgi?id=297891</a>

Reviewed by Sergio Villar Senin.

We should move the file descriptors in makeExternalImageSource instead.

Canonical link: <a href="https://commits.webkit.org/299152@main">https://commits.webkit.org/299152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43aeecce73a8d485f12111224e4175181972d137

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118040 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28354 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124190 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119918 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46300 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89575 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59176 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105844 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70067 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29662 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23961 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67855 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100017 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24139 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127271 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33865 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98247 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45304 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102068 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98033 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43433 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21423 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41379 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18812 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44815 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50489 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44275 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47620 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45964 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar reviewed (approved):
I was a bit scared to do the move there due to potential missuses of that field after that but I guess it's fine, I don't foresee any future changes that could require access to the ExternalTexture fds

carlosgcampos committed:
[OpenXR] Do not duplicate the DMA-BUF file descriptors before importing the external texture
https://bugs.webkit.org/show_bug.cgi?id=297891

Reviewed by Sergio Villar Senin.

We should move the file descriptors in makeExternalImageSource instead.

Canonical link: https://commits.webkit.org/299152@main

spenap commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -71,14 +71,12 @@ static void createAndBindCompositorBuffer(GL& gl, WebXRExternalRenderbuffer& buf
-        }),
-        .strides = imageSource.strides,
-        .offsets = imageSource.offsets,
+        .fds = WTFMove(imageSource.fds),
+        .strides = WTFMove(imageSource.strides),
+        .offsets = WTFMove(imageSource.offsets),
```
While this moves instead of duplicating, is it OK from the point of view of the caller? 

Are callers passing in an `imageSource` prepare to handle moved-out resource? Because we're not doing exchange: the old values are now in a valid, unspecified value if I understand it correctly.

carlosgcampos replied:
Yes, we just check if the fds vector is empty to decide and it's empty after the move

carlosgcampos replied:
Because in the case of Vector move always makes the moved one empty.
